Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +133,757 @@ final class ConnectionPool {
+                    // this connections is now gone, we will either create new connection or do nothing
+                    self.openedConnectionsCount -= 1
+                    self.leasedConnections.remove(connection)
+
+                    return self.processNextWaiter()
                 }
+            case .closed:
+                self.openedConnectionsCount -= 1
+                self.leasedConnections.remove(connection)
 
-                connection.closePromise.succeed(())
+                return self.processNextWaiter()
             }
         }
 
-        /// Removes and fails all `waiters`, remove existing `availableConnections` and sets `state.activity` to `.closing`
-        func prepareForClose() -> EventLoopFuture<Void> {
-            let (waitersFutures, closeFutures) = self.stateLock.withLock { () -> ([EventLoopFuture<Connection>], [EventLoopFuture<Void>]) in
-                // Fail waiters
-                let waitersCopy = self.state.waiters
-                self.state.waiters.removeAll()
-                let waitersPromises = waitersCopy.map { $0.promise }
-                let waitersFutures = waitersPromises.map { $0.futureResult }
-                waitersPromises.forEach { $0.fail(HTTPClientError.cancelled) }
-                let closeFutures = self.state.availableConnections.map { $0.close() }
-                return (waitersFutures, closeFutures)
+        mutating func offer(connection: Connection) -> Action {
+            switch self.state {
+            case .active:
+                self.leasedConnections.insert(connection)
+                return .none
+            case .closed: // This can happen when we close the client while connections was being estableshed
+                return .cancel(connection, self.isEmpty)
             }
-
-            return EventLoopFuture<Connection>.andAllComplete(waitersFutures, on: self.eventLoop)
-                .flatMap {
-                    EventLoopFuture<Void>.andAllComplete(closeFutures, on: self.eventLoop)
-                }
-                .map { _ in
-                    self.stateLock.withLock {
-                        if self.state.leased == 0, self.state.availableConnections.isEmpty {
-                            self.state.activity = .closed
-                        } else {
-                            self.state.activity = .closing
-                        }
-                    }
-                }
         }
 
-        func close() -> EventLoopFuture<Void> {
-            let availableConnections = self.stateLock.withLock { () -> CircularBuffer<ConnectionPool.Connection> in
-                assert(self.state.activity == .closing)
-                return self.state.availableConnections
+        mutating func drop(connection: Connection) {
+            switch self.state {
+            case .active:
+                self.leasedConnections.remove(connection)
+            default:
```

Comment:
I think we should remove all `default:`s in the state machine, that way you get compile errors anywhere you need to fix stuff when you add a case. [Here](https://github.com/apple/swift-nio-examples/blob/master/backpressure-file-io-channel/Sources/BackpressureChannelToFileIO/FileIOChannelWriteCoordinator.swift#L265-L275)'s an example of who this can still be very concise.

I'd also recommend adding a function like

```
    private func illegalTransition(_ function: String = #function) -> Never {
        preconditionFailure("illegal transition \(function) in \(self)")
    }
```

to get really good error messages. The function name is the "input" to the state machine, therefore you get exactly what you need to debug: The current state + the input :)

Reply:
done